tools/fs/safe-watcher.ts

```diff
@@ -1,476 +1,364 @@
-
-  function watchWrapper(event: string) {
-    lastWatcherEventTime = Date.now();
-    fire(event);
-
-    // It's tempting to call unwatchFile(absPath, watchFileWrapper) here,
-    // but previous watcher success is no guarantee of future watcher
-    // reliability. For example, watchLibrary.watch works just fine when file
-    // changes originate from within a Vagrant VM, but changes to shared
-    // files made outside the VM are invisible to watcher, so our only
-    // hope of catching them is to continue polling.
   }
 
-  function rewatch() {
-    if (hasPriority(absPath)) {
-      if (watcher) {
-        // Already watching; nothing to do.
-        return;
+  // Remove any existing roots that are now encompassed by the new one.
+  for (const root of Array.from(watchRoots)) {
+    const rel = pathRelative(dirPath, root);
+    if (root !== dirPath && !rel.startsWith("..") && !rel.startsWith("/")) {
+      const sub = dirSubscriptions.get(root);
+      if (sub) {
+        try {
+          await sub.unsubscribe();
+        } catch (_) {
+          /* ignore errors */
+        }
       }
-      watcher = watchLibraryWatch(absPath, watchWrapper);
-    } else if (watcher) {
-      safeUnwatch();
+      dirSubscriptions.delete(root);
+      watchRoots.delete(root);
     }
-
-    // Since we're about to restart the stat-based file watcher, we don't
-    // want to miss any of its events because of the lastWatcherEventTime
-    // throttling that it attempts to do.
-    lastWatcherEventTime = 0;
-
-    // We use files.watchFile in addition to watcher.watch as a fail-safe
-    // to detect file changes even on network file systems.  However
-    // (unless the user disabled watcher or this watcher call failed), we
-    // use a relatively long default polling interval of 5000ms to save
-    // CPU cycles.
-    statWatch(absPath, getPollingInterval(), watchFileWrapper);
   }
 
-  function watchFileWrapper(newStat: Stats, oldStat: Stats) {
-    if (newStat.ino === 0 &&
-        oldStat.ino === 0 &&
-        +newStat.mtime === +oldStat.mtime) {
-      // Node calls the watchFile listener once with bogus identical stat
-      // objects, which should not trigger a file change event.
+  const osDirPath = convertToOSPath(dirPath);
+  // Check that osDirPath is indeed a directory.
+  try {
+    const stats = statOrNull(osDirPath);
+    if (!stats?.isDirectory()) {
+      console.warn(`Skipping watcher for ${osDirPath}: not a directory`);
+      ignoredWatchRoots.add(dirPath);
       return;
     }
+  } catch (e) {
+    console.error(`Failed to stat ${osDirPath}:`, e);
+    ignoredWatchRoots.add(dirPath);
+    return;
+  }
 
-    // If a watcher event fired in the last polling interval, ignore
-    // this event.
-    if (Date.now() - lastWatcherEventTime > getPollingInterval()) {
-      fire("change");
+  // Set up ignore patterns to skip deep node_modules and .meteor/local cache
+  const ignorePatterns = ["**/node_modules/**", "**/.meteor/local/**"];
```

Comment:
I'm ignoring `node_modules` and `.meteor/local` in `ParcelWatcher` and with `shouldIgnorePath`, since in a new app, changes there don't trigger the Meteor bundler. I didn’t find this exclusion in the previous safe-watcher, but without it, profile times increase a lot.